### ISSUE TITLE
Warn when functions that exit are invoked in subshells

### ISSUE THREAD

**Issue by koalaman** [koalaman/shellcheck#1558]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

Inspired by #1158

```sh
#!/bin/sh

my_mktemp() {
  mktemp || exit 42
}

dir=$(my_mktemp)
echo "$dir"
```

#### Here's what shellcheck currently says:

Nothing

#### Here's what I wanted or expected to see:

The `exit` is buggy here, since the function is only ever called in a subshell so it'll never do anything other than `return`
